torchmultimodal/utils/generate.py

```diff
@@ -0,0 +1,314 @@
+            idx += 1
+
+        samples = torch.cat(
+            samples[1:], dim=1
+        )  # remove the "sos" token: (b, out_seq_len)
+        samples = self.model.decode(samples)  # type: ignore
+
+        return SampleOutput(samples=samples, model_outputs=model_outputs)
+
+    def _filter_logits(
+        self, logits: Tensor, top_k: Optional[int] = None, top_p: Optional[float] = None
+    ) -> Tensor:
+        logits_filters: List[Any] = []
+        if top_k is not None:
+            logits_filters.append(LogitsFilterTopK(top_k))
+        if top_p is not None:
+            logits_filters.append(LogitsFilterTopP(top_p))
+
+        for _filter in logits_filters:
+            logits = _filter(logits)
+
+        return logits
+
+
+def get_logits_mask(
+    in_seq_len: int = 0,
+    out_seq_len: int = 0,
+    num_in_tokens: int = 0,
+    num_out_tokens: int = 0,
+) -> Tensor:
+    """Applies masks to logits to restrict prediction from being made of tokens of the opposite modality.
+
+    Args:
+        in_seq_len (int, optional): Length of input modality sequence from the logits tensor. Defaults to ``0``.
+        out_seq_len (int, optional): Length of output modality sequence from the logits tensor.
+            Defaults to ``0``.
+        num_in_tokens (int, optional): Number of input modality token states from the model. Defaults to ``0``.
+        num_out_tokens (int, optional): Number of output modality token states from the model.
+            Defaults to ``0``.
+
+    Returns:
+        Logits mask tensor containing ``1``s for unmasked positions and ``0``s for masked ones.
+    """
+    mask = torch.zeros(in_seq_len + out_seq_len, num_in_tokens + num_out_tokens)
+    # the quadrant of input modality sequence, input token states should not be masked
+    # Similar for that of output modality sequence, output token states
+    mask[in_seq_len:, num_in_tokens:] = 1
+    mask[:in_seq_len, :num_in_tokens] = 1
+
+    return mask
+
+
+class LogitsFilterTopK:
+    """Filters a distribution of logits using top_k
+
+    Attributes:
+        top_k (int, optional): Keeps the top_k tokens with the highest probability (top_k filtering).
+            Defaults to ``None``.
+        filter_value (float, optional): Constant value to filter unwanted logits. Defaults to ``-inf``.
+        min_tokens_to_keep (int, optional): Minimum number of tokens to keep per batch example in the output.
+            Defaults to ``1``.
+
+    Code reference: https://gist.github.com/thomwolf/1a5a29f6962089e871b94cbd09daf317
+
+    Args:
+        logits (Tensor): Logits distribution shape ``(b, num_tokens)`` where ``b`` is batch size,
+            ``num_tokens`` is the number of tokens.
+
+    Returns:
+        Filtered logits tensor.
+
+    Raises:
+        ValueError: If 'top_k' is outside of valid numerical ranges.
+
+    """
+
+    def __init__(
+        self,
+        top_k: Optional[int] = None,
+        min_tokens_to_keep: int = 1,
```

Comment:
Just an FYI that as written this param is not exposed in the `GenerationUtil` class. Probably not a huge deal, but just wanna point it out in case you think people would want to configure this from the main class

Reply:
Thanks. I suspect `min_tokens_to_keep` is not the first order param to be exposed. For example, in Huggingface this param is controlled by `num_beams` (for beam search):

https://github.com/huggingface/transformers/blob/v4.21.2/src/transformers/generation_utils.py#L663

Until then we could probably keep it "hidden" from the user (just like `filter_value` being `-float("inf")`.)